### PR TITLE
xtensa: mask interrupts earlier

### LINKED ISSUE #16915: stack_sentinel: rare ASSERTION FAIL [!(z_arch_curr_cpu()->nested != 0U)] @ ZEPHYR_BASE/kernel/thread.c:429  Threads may not be created in ISRs
**Describe the bug**
Intermittent failure of sanitycheck test:
```
qemu_xtensa               tests/kernel/fatal/kernel.common.stack_sentinel
```
Spotted only once. Generally doesn't fail. The previous run was in the exact same conditions and didn't fail.

**To Reproduce**
1. Run this test in sanitycheck many times. On a loaded machine maybe?
2. Pray

**Expected behavior**
PASS

**Impact**
Low.

**Environment (please complete the following information):**
 - OS: Ubuntu 18.0.4.2
 - Toolchain Zephyr SDK 0.10.0
 - Commit SHA or Version used: b58aa20e135a

**Additional context**
JOBS=50

**Screenshots or console output**

```
qemu_xtensa               tests/kernel/fatal/kernel.common.stack_sentinel    FAILED: unexpected byte (qemu)`
***** Booting Zephyr build zephyr-v1.14.0-1527-g565bf3263d3b *****
Running test suite fatal
===================================================================
starting test - test_fatal
test alt thread 1: generic CPU exception
 ** FATAL EXCEPTION
 ** CPU 0 EXCCAUSE 0 PS 0x00060020 PC 0x600044c0 VADDR 0x00000000
 **  A0 0xa0001a72  SP 0x600044c0  A2 0x00000000  A3 0x00000000
 **  A4 0x00000000  A5 0x60004b10  A6 0x00000000  A7 0x00060520
 **  A8 0xa0000f68  A9 0x00000040 A10 0x00000000 A11 0x60004500
 ** A12 0x60004b10 A13 0x60004af0 A14 0xfffffff5 A15 0x00060520
 ** SAR 0x00000000
Current thread ID = 0x60006088
Faulting instruction address = 0xdeaddead
Caught system error -- reason 0
test alt thread 2: initiate kernel oops
ASSERTION FAIL [!(z_arch_curr_cpu()->nested != 0U)] @ ZEPHYR_BASE/kernel/thread.c:429
        Threads may not be created in ISRs
@ ZEPHYR_BASE/lib/os/assert.c:30:
***** Kernel Panic! *****
Current thread ID = 0x60006104
Faulting instruction address = 0xdeaddead
Caught system error -- reason 6

    Assertion failed at ZEPHYR_BASE/tests/kernel/fatal/src/main.c:247: test_fatal: (crash_reason not equal to _NANO_ERR_KERNEL_OOPS)
bad reason code got 6 expected 5

test alt thread 3: initiate kernel panic
ASSERTION FAIL [!(z_arch_curr_cpu()->nested != 0U)] @ ZEPHYR_BASE/kernel/thread.c:429
        Threads may not be created in ISRs
@ ZEPHYR_BASE/lib/os/assert.c:30:
***** Kernel Panic! *****
Current thread ID = 0x60006104
Faulting instruction address = 0xdeaddead
Caught system error -- reason 6
test stack sentinel overflow - timer irq
ASSERTION FAIL [!(z_arch_curr_cpu()->nested != 0U)] @ ZEPHYR_BASE/kernel/thread.c:429
        Threads may not be created in ISRs
@ ZEPHYR_BASE/lib/os/assert.c:30:
***** Kernel Panic! *****
Current thread ID = 0x60006104
Faulting instruction address = 0xdeaddead
Caught system error -- reason 6

    Assertion failed at ZEPHYR_BASE/tests/kernel/fatal/src/main.c:197: check_stack_overflow: (crash_reason not equal to _NANO_ERR_STACK_CHK_FAIL)
bad reason code got 6 expected 2

FAIL - test_fatal
===================================================================
Test suite fatal failed.
===================================================================
PROJECT EXECUTION FAILED
```

### PR DESCRIPTION
**Pull request by tyrantlizard77**
When coming out of an exception, we need to mask interrupts
to avoid races when decrementing the nested count. Move
the instruction that does this earlier.

Fixes: #16915 

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

**Comment by andrewboie** [comment]
This is a theoretical fix, based on code inspection and discussion with @andyross. I could not get the issue to reproduce locally.

**Review by ioannisg — COMMENTED** [review]
Makes sense for me.
I leave this to the xtensa maintainers for formal approval

**Comment by ioannisg** [comment]
@dcpleung could you provide some feedback here?

**Comment by dcpleung** [comment]
Register a0 will be overwritten at line 296, though I don't see it being used later.

@andyross should be able to definite say whether this is correct or not, since it's his code.

**Review by andyross — APPROVED** [review]
Yeah, that's a race for sure.  The handlers will drop the exception level to whatever is appropriate for their level, and nothing was ever increasing it.  So a nested interrupt that landed at the wrong space could see an incorrect nest count.

The change in the value of a0 exiting the sequence is irrelevant (it used to be left as the nest offset, now it's the old exception level).  We're about to do a full restore at this stage anyway, so nothing is going to care what value these registers hold.

**Pull request merged**
